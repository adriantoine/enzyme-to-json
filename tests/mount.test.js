/* eslint-env jest */

import React from 'react';
import Enzyme, {mount} from 'enzyme';
import omitBy from 'lodash/omitBy';
import Adapter from 'enzyme-adapter-react-16';

import mountToJson from '../src/mount';
import {
  BasicPure,
  BasicWithUndefined,
  BasicWithAList,
  ComponentWithAZeroChildren,
  ArrayRender,
  FalsyTruthyComponent,
  FalsyChildren,
  FragmentAsChild,
  FragmentAsRoot,
} from './fixtures/pure-function';
import {
  BasicClass,
  ClassWithPure,
  ClassWithDirectPure,
  ClassWithDirectComponent,
  ClassWithNull,
  ClassArrayRender,
} from './fixtures/class';

Enzyme.configure({adapter: new Adapter()});

function WrapperComponent(props) {
  return <BasicPure {...props} />;
}

it('doesnt break when called without arguments', () => {
  expect(mountToJson()).toBe(null);
  expect(mountToJson([null])).toBe(null);
  expect(mountToJson(['a'])).toBe(null);
});

it('converts basic pure mount', () => {
  const mounted = mount(
    <BasicPure className="pure">
      <strong>Hello!</strong>
    </BasicPure>,
  );

  expect(mountToJson(mounted)).toMatchSnapshot();
});

it('converts pure mount with mixed children', () => {
  const mounted = mount(<BasicPure>Hello {'world'}!</BasicPure>);
  expect(mountToJson(mounted)).toMatchSnapshot();
});

it('converts basic class mount', () => {
  const mounted = mount(
    <BasicClass className="class">
      <strong>Hello!</strong>
    </BasicClass>,
  );

  expect(mountToJson(mounted)).toMatchSnapshot();
});

it('converts class mount with mixed children', () => {
  const mounted = mount(<BasicClass>Hello {'world'}!</BasicClass>);
  expect(mountToJson(mounted)).toMatchSnapshot();
});

it('converts a class mount with a pure function in it', () => {
  const mounted = mount(
    <ClassWithPure className="class">
      <strong>Hello!</strong>
    </ClassWithPure>,
  );

  expect(mountToJson(mounted)).toMatchSnapshot();
});

it('converts a class mount with a pure function in it as a direct child', () => {
  const mounted = mount(
    <ClassWithDirectPure className="class">
      <strong>Hello!</strong>
    </ClassWithDirectPure>,
  );

  expect(mountToJson(mounted)).toMatchSnapshot();
});

it('converts a class mount with a class component in it as a direct child', () => {
  const mounted = mount(
    <ClassWithDirectComponent className="class">
      <strong>Hello!</strong>
    </ClassWithDirectComponent>,
  );

  expect(mountToJson(mounted)).toMatchSnapshot();
});

it('handles a component which returns null', () => {
  const mounted = mount(<ClassWithNull />);

  expect(mountToJson(mounted)).toMatchSnapshot();
});

it('skips undefined props', () => {
  const mounted = mount(<BasicWithUndefined>Hello!</BasicWithUndefined>);

  expect(mountToJson(mounted)).toMatchSnapshot();
});

it('renders zero-children', () => {
  const mounted = mount(<ComponentWithAZeroChildren />);
  expect(mountToJson(mounted)).toMatchSnapshot();
});

it('renders multiple elements as a result of find', () => {
  const mounted = mount(<BasicWithAList />);
  expect(mountToJson(mounted.find('li'))).toMatchSnapshot();
});

it('outputs the key prop', () => {
  const mounted = mount(
    <ul>
      <li key={1} />
      <li key={2} />
    </ul>,
  );
  expect(mountToJson(mounted)).toMatchSnapshot();
});

it('doesnt output the key prop when noKey option is passed', () => {
  const mounted = mount(
    <ul>
      <li key={1} />
      <li key={2} />
    </ul>,
  );
  expect(mountToJson(mounted, {noKey: true})).toMatchSnapshot();
});

it('converts function components with render returning top level arrays', () => {
  const mounted = mount(
    <ArrayRender>
      <strong>Hello!</strong>
    </ArrayRender>,
  );

  expect(mountToJson(mounted)).toMatchSnapshot();
});

it('converts class components with render returning top level arrays', () => {
  const mounted = mount(
    <ClassArrayRender>
      <strong>Hello!</strong>
    </ClassArrayRender>,
  );

  expect(mountToJson(mounted)).toMatchSnapshot();
});

it('handles elements in prop arrays', () => {
  const mounted = mount(
    <WrapperComponent
      elements={[
        <BasicPure>
          <strong>Hello!</strong>
        </BasicPure>,
      ]}
    />,
  );

  expect(mountToJson(mounted)).toMatchSnapshot();
});

it('handles elements in prop objects', () => {
  const mounted = mount(
    <WrapperComponent
      element={{
        element: (
          <BasicPure>
            <strong>Hello!</strong>
          </BasicPure>
        ),
        nestedElements: [
          <BasicPure>
            <strong>Hello again!</strong>
          </BasicPure>,
        ],
      }}
    />,
  );

  expect(mountToJson(mounted)).toMatchSnapshot();
});

it('accepts a map option allowing to customize content', () => {
  const mounted = mount(<strong>Hello!</strong>);

  expect(
    mountToJson(mounted, {
      map: json => ({...json, children: ['Goodbye!']}),
    }),
  ).toMatchSnapshot();
});

it('accepts a map option allowing to customize content of all nested components', () => {
  const mounted = mount(
    <div randomlygeneratedkey={Date.now()} className="wrapper">
      <strong randomlygeneratedkey={Date.now()}>Hello!</strong>
      <strong className="strong2">Hello 2</strong>
    </div>,
  );

  expect(
    mountToJson(mounted, {
      map: json => ({
        ...json,
        props: omitBy(json.props, (val, key) => key === 'randomlygeneratedkey'),
      }),
    }),
  ).toMatchSnapshot();
});

it('can skip a component I dont want to see with the map option', () => {
  const mounted = mount(
    <div>
      <ul>
        <li>Item 1</li>
        <li>Item 2</li>
      </ul>
      <strong>Hello 2</strong>
    </div>,
  );

  expect(
    mountToJson(mounted, {
      map: json => {
        if (json.type === 'ul') {
          return null;
        }
        return json;
      },
    }),
  ).toMatchSnapshot();
});

it('outputs the snapshot even with inline JSX conditions being falsy', () => {
  const mounted = mount(
    <div>
      <span>I am there</span>
      {false && <span>Issue</span>}
    </div>,
  );

  expect(mountToJson(mounted)).toMatchSnapshot();
});

it('outputs an empty string when a component returns false', () => {
  const mounted = mount(<FalsyTruthyComponent foo={false} />);
  expect(mountToJson(mounted)).toMatchSnapshot();
});

it('outputs an empty string when a component has false chidren', () => {
  const mounted = mount(<FalsyChildren falsy={false} />);
  expect(mountToJson(mounted)).toMatchSnapshot();
});

it('renders a component that has a child fragment', () => {
  const wrapper = mount(<FragmentAsChild />);

  expect(mountToJson(wrapper)).toMatchSnapshot();
});

it('renders a component that has a fragment root', () => {
  const wrapper = mount(<FragmentAsRoot />);

  expect(mountToJson(wrapper)).toMatchSnapshot();
});
