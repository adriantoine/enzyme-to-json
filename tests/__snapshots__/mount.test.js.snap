// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accepts a map option allowing to customize content 1`] = `
<strong>
  Goodbye!
</strong>
`;

exports[`accepts a map option allowing to customize content of all nested components 1`] = `
<div
  className="wrapper"
>
  <strong>
    Hello!
  </strong>
  <strong
    className="strong2"
  >
    Hello 2
  </strong>
</div>
`;

exports[`can skip a component I dont want to see with the map option 1`] = `
<div>
  <strong>
    Hello 2
  </strong>
</div>
`;

exports[`converts a class mount with a class component in it as a direct child 1`] = `
<ClassWithDirectComponent
  className="class"
>
  <ClassWithPure
    className="nested-pure"
  >
    <div
      className="class-with-pure nested-pure"
      onClick={[Function]}
    >
      <BasicPure
        className="nested-pure"
      >
        <div
          className="basic-pure nested-pure"
          onClick={[Function]}
        >
          <div
            className="group"
            id="group-id"
          >
            <span>
              <span>
                <span>
                  <strong>
                    Hello!
                  </strong>
                </span>
                <span
                  className="empty"
                />
              </span>
              <span
                className="empty"
              />
            </span>
            <span
              className="empty"
            />
          </div>
        </div>
      </BasicPure>
    </div>
  </ClassWithPure>
</ClassWithDirectComponent>
`;

exports[`converts a class mount with a pure function in it 1`] = `
<ClassWithPure
  className="class"
>
  <div
    className="class-with-pure class"
    onClick={[Function]}
  >
    <BasicPure
      className="nested-pure"
    >
      <div
        className="basic-pure nested-pure"
        onClick={[Function]}
      >
        <div
          className="group"
          id="group-id"
        >
          <span>
            <span>
              <strong>
                Hello!
              </strong>
            </span>
            <span
              className="empty"
            />
          </span>
          <span
            className="empty"
          />
        </div>
      </div>
    </BasicPure>
  </div>
</ClassWithPure>
`;

exports[`converts a class mount with a pure function in it as a direct child 1`] = `
<ClassWithDirectPure
  className="class"
>
  <BasicPure
    className="nested-pure"
  >
    <div
      className="basic-pure nested-pure"
      onClick={[Function]}
    >
      <div
        className="group"
        id="group-id"
      >
        <span>
          <span>
            <strong>
              Hello!
            </strong>
          </span>
          <span
            className="empty"
          />
        </span>
        <span
          className="empty"
        />
      </div>
    </div>
  </BasicPure>
</ClassWithDirectPure>
`;

exports[`converts basic class mount 1`] = `
<BasicClass
  className="class"
>
  <div
    className="basic-class class"
    onClick={[Function]}
  >
    <div
      className="group"
      id="group-id"
    >
      <span>
        <strong>
          Hello!
        </strong>
      </span>
      <span
        className="empty"
      />
    </div>
  </div>
</BasicClass>
`;

exports[`converts basic pure mount 1`] = `
<BasicPure
  className="pure"
>
  <div
    className="basic-pure pure"
    onClick={[Function]}
  >
    <div
      className="group"
      id="group-id"
    >
      <span>
        <strong>
          Hello!
        </strong>
      </span>
      <span
        className="empty"
      />
    </div>
  </div>
</BasicPure>
`;

exports[`converts class components with render returning top level arrays 1`] = `
<ClassArrayRender>
  <div
    className="test"
    key="test"
  >
    Test
  </div>
  <div
    className="test2"
    key="test2"
  >
    Test 2
  </div>
  <div
    className="child"
    key="child"
  >
    <strong>
      Hello!
    </strong>
  </div>
</ClassArrayRender>
`;

exports[`converts class mount with mixed children 1`] = `
<BasicClass>
  <div
    className="basic-class undefined"
    onClick={[Function]}
  >
    <div
      className="group"
      id="group-id"
    >
      <span>
        Hello 
        world
        !
      </span>
      <span
        className="empty"
      />
    </div>
  </div>
</BasicClass>
`;

exports[`converts function components with render returning top level arrays 1`] = `
<ArrayRender>
  <div
    className="test"
    key="test"
  >
    Test
  </div>
  <div
    className="test2"
    key="test2"
  >
    Test 2
  </div>
  <div
    className="child"
    key="child"
  >
    <strong>
      Hello!
    </strong>
  </div>
</ArrayRender>
`;

exports[`converts pure mount with mixed children 1`] = `
<BasicPure>
  <div
    className="basic-pure undefined"
    onClick={[Function]}
  >
    <div
      className="group"
      id="group-id"
    >
      <span>
        Hello 
        world
        !
      </span>
      <span
        className="empty"
      />
    </div>
  </div>
</BasicPure>
`;

exports[`doesnt output the key prop when noKey option is passed 1`] = `
<ul>
  <li />
  <li />
</ul>
`;

exports[`handles a component which returns null 1`] = `<ClassWithNull />`;

exports[`handles elements in prop arrays 1`] = `
<WrapperComponent
  elements={
    Array [
      <BasicPure>
        <strong>
          Hello!
        </strong>
      </BasicPure>,
    ]
  }
>
  <BasicPure
    elements={
      Array [
        <BasicPure>
          <strong>
            Hello!
          </strong>
        </BasicPure>,
      ]
    }
  >
    <div
      className="basic-pure undefined"
      onClick={[Function]}
    >
      <div
        className="group"
        id="group-id"
      >
        <span />
        <span
          className="empty"
        />
      </div>
    </div>
  </BasicPure>
</WrapperComponent>
`;

exports[`handles elements in prop objects 1`] = `
<WrapperComponent
  element={
    Object {
      "element": <BasicPure>
        <strong>
          Hello!
        </strong>
      </BasicPure>,
      "nestedElements": Array [
        <BasicPure>
          <strong>
            Hello again!
          </strong>
        </BasicPure>,
      ],
    }
  }
>
  <BasicPure
    element={
      Object {
        "element": <BasicPure>
          <strong>
            Hello!
          </strong>
        </BasicPure>,
        "nestedElements": Array [
          <BasicPure>
            <strong>
              Hello again!
            </strong>
          </BasicPure>,
        ],
      }
    }
  >
    <div
      className="basic-pure undefined"
      onClick={[Function]}
    >
      <div
        className="group"
        id="group-id"
      >
        <span />
        <span
          className="empty"
        />
      </div>
    </div>
  </BasicPure>
</WrapperComponent>
`;

exports[`outputs an empty string when a component has false chidren 1`] = `
<FalsyChildren
  falsy={false}
>
  <div />
</FalsyChildren>
`;

exports[`outputs an empty string when a component returns false 1`] = `
<FalsyTruthyComponent
  foo={false}
/>
`;

exports[`outputs the key prop 1`] = `
<ul>
  <li
    key="1"
  />
  <li
    key="2"
  />
</ul>
`;

exports[`outputs the snapshot even with inline JSX conditions being falsy 1`] = `
<div>
  <span>
    I am there
  </span>
</div>
`;

exports[`renders a component that has a child fragment 1`] = `
<FragmentAsChild>
  <div>
    <span />
    <div />
    <button />
  </div>
</FragmentAsChild>
`;

exports[`renders a component that has a fragment root 1`] = `
<FragmentAsRoot>
  <span />
  <div />
  <button />
</FragmentAsRoot>
`;

exports[`renders multiple elements as a result of find 1`] = `
Array [
  <li>
    0
  </li>,
  <li>
    1
  </li>,
]
`;

exports[`renders zero-children 1`] = `
<ComponentWithAZeroChildren>
  <div>
    0
  </div>
</ComponentWithAZeroChildren>
`;

exports[`skips undefined props 1`] = `
<BasicWithUndefined>
  <button>
    Hello
  </button>
</BasicWithUndefined>
`;
